src/compiler/checker.ts

```diff
@@ -8884,6 +8896,17 @@ namespace ts {
         function getResolvedMembersOrExportsOfSymbol(symbol: Symbol, resolutionKind: MembersOrExportsResolutionKind): UnderscoreEscapedMap<Symbol> {
             const links = getSymbolLinks(symbol);
             if (!links[resolutionKind]) {
+                // We _must_ resolve any possible commonjs export assignments _before_ merging in late bound members, as cjs export assignments may cause
+                // things to be merged (?!?) into this symbol; moreover, we _can't_ resolve those export assignments if we're already resolving the containing

```

Comment:
```js
function Foo() {
}
const lateBound = "field";
Foo[lateBound] = 12;
Foo.A = function() {}
/**
 * @typedef {number | string} FooProp
 */
module.exports = Foo.A;
```
The export assignment is the A prop of food, which means we need the full list of Foo's exports to check if A is a member (and what it is), which means we need to do a cjs merge to merge the typedefs into the exports, which means we need to know the target of Foo.A, which means we need the A member of Foo, which means we need Foo's exports, which means....

Yeah, the process is inherently circular. This peek short circuits it, since symbol merging makes all the partial results layer together in the end anyway.

Reply:
> which means we need to do a cjs merge to merge the typedefs into the exports

I’m missing how we get to here... did you mean to _use_ `FooProp` in your example scenario? I can’t tell why resolving anything related to `Foo` would require resolving any exports of the top-level module.

Reply:
> I’m missing how we get to here... did you mean to use FooProp in your example scenario?

Nope. I did not.

> I can’t tell why resolving anything related to Foo would require resolving any exports of the top-level module.

The "top level module" is defined by a `export=` member which is an alias pointing at `Foo.A`. (And a typedef). We don't _really_ support having both an `export=` and another top-level export, so to handle that scenario, we _merge_ the other top-level export into the target of the `export=` alias symbol. I suppose the more telling example is when the final line is `module.exports = Foo;` instead. `Foo` clearly needs to have both `FooProp` and `"field"` members merged into it. OK, so you handle that by layering the export table merges (and forcing the cjs exports merge to always come first) - once that's done, you look at the `module.exports = Foo.A` case. Well, in resolving `Foo.A`, you need the exports of `Foo`, which means getting its exports; before it's safe to get its exports, we _don't yet know_ if it's the target of the `export =` (imagine if `Foo.A = Foo`!), so we must perform the containing file's cjs merge (to ensure the exports of the target are complete, in case the target winds up being the symbol we're about to look up). Doing so would attempt to resolve `export=` again, which is a circularity, as we are already doing that.

Reply:
> we _merge_ the other top-level export into the target of the export= alias symbol

Ohhh, yep, I get it now. Thanks 👍